src/main/rx/srxl.c

```diff
@@ -0,0 +1,215 @@
+#include "io/serial.h"
+
+#include "rx/rx.h"
+#include "rx/srxl.h"
+
+#include "common/crc.h"
+
+
+
+#define SRXL_CHANNEL_COUNT_A1 12
+#define SRXL_CHANNEL_COUNT_A2 16
+
+#define SRXL_CHANNEL_COUNT_MAX SRXL_CHANNEL_COUNT_A2
+
+
+
+// Frame is: ID(1 byte) + 12*channel(2 bytes) + CRC(2 bytes) = 27
+#define SRXL_FRAME_SIZE_A1 (SRXL_CHANNEL_COUNT_A1*2 + 3)
+// Frame is: ID(1 byte) + 16*channel(2 bytes) + CRC(2 bytes) = 35
+#define SRXL_FRAME_SIZE_A2 (SRXL_CHANNEL_COUNT_A2*2 + 3)
+
+#define SRXL_BAUDRATE 115200
+#define SRXL_MAX_FRAME_TIME 8000
+
+// NOTE!
+// This is actually based on ID+LENGTH (nibble each)
+// 0xA - Multiplex ID (also used by JR, no idea why)
+// 0x1 - 12 channels
+// 0x2 - 16 channels
+// Add identifier here if another manufacturer uses another ID
+
+#define SRXL_START_OF_FRAME_BYTE_A1 (0xA1)		//12 channels
+#define SRXL_START_OF_FRAME_BYTE_A2 (0xA2)		//16 channels transfer
+
+// Pulse length convertion from [0...4095] to µs:
+//      800µs  -> 0x000
+//      1500µs -> 0x800
+//      2200µs -> 0xFFF
+// Total range is: 2200 - 800 = 1400 <==> 4095
+// Use formula: 800 + value * 1400 / 4096 (i.e. a shift by 12)
+#define SRXL_CONVERT_TO_USEC(V)	(800 + ((V * 1400) >> 12))
+
+static bool srxlFrameReceived = false;
+static bool srxlDataIncoming = false;
+static uint8_t srxlFramePosition;
+static uint8_t srxlFrameLength;
+static uint8_t srxlChannelCount;
+
+static void srxlDataReceive(uint16_t c);
+
+// Use max values for ram areas
+static volatile uint8_t srxlFrame[SRXL_FRAME_SIZE_A2];	//size 35 for 16 channels in SRXL 0xA2
+static uint16_t srxlChannelData[SRXL_CHANNEL_COUNT_MAX];
+
+uint8_t srxlCRC_OK();
+
+static uint16_t srxlReadRawRC(rxRuntimeConfig_t *rxRuntimeConfig, uint8_t chan);
+
+bool srxlInit(rxRuntimeConfig_t *rxRuntimeConfig, rcReadRawDataPtr *callback)
+{
+    uint32_t baudRate;
+    rxRuntimeConfig->channelCount = SRXL_CHANNEL_COUNT_MAX;
+    srxlFrameReceived = false;
+    srxlDataIncoming = false;
+    srxlFramePosition = 0;
+    baudRate = SRXL_BAUDRATE;
+    srxlFrameLength = SRXL_FRAME_SIZE_A2;		//default for 12 channel
+    srxlChannelCount = SRXL_CHANNEL_COUNT_MAX;
+
+    if (callback) {
+        *callback = srxlReadRawRC;
+    }
+
+    serialPortConfig_t *portConfig = findSerialPortConfig(FUNCTION_RX_SERIAL);
+    if (!portConfig) {
+        return false;
+    }
+
+    serialPort_t *srxlPort = openSerialPort(portConfig->identifier, FUNCTION_RX_SERIAL, srxlDataReceive, baudRate, MODE_RX, SERIAL_NOT_INVERTED);
+
+    return srxlPort != NULL;
+}
+
+
+uint8_t srxlCRC_OK()
+{
+	uint8_t i = 0;
+    // Calculate the CRC of the incoming frame
+    uint16_t crc = 0;
+    uint16_t inCrc = 0;
+
+    // Calculate on all bytes except the final two CRC bytes
+    for (i = 0; i < srxlFrameLength - 2; i++) {
+        inCrc =  crc16_CCITT(inCrc, srxlFrame[i]);
+    }
+
+    // Get the received CRC
+    crc = ((uint16_t)srxlFrame[srxlFrameLength - 2]) << 8;
+    crc = crc + ((uint16_t)srxlFrame[srxlFrameLength - 1]);
```

Comment:
Its easier to compute CRC of whole message (including CRC itself), resulting value will be  constant for correct message (zero when using 0 as starting value)